tree/ntuple/v7/src/RClusterPool.cxx

```diff
@@ -0,0 +1,329 @@
+         auto existingCluster = FindInPool(cptr->GetId());
+         if (existingCluster) {
+            existingCluster->Adopt(std::move(*cptr));
+         } else {
+            auto idxFreeSlot = FindFreeSlot();
+            fPool[idxFreeSlot] = std::move(cptr);
+         }
+         itr = fInFlightClusters.erase(itr);
+      }
+
+      // Determine clusters which get triggered for background loading
+      for (auto &cptr : fPool) {
+         if (!cptr)
+            continue;
+         provide.Erase(cptr->GetId(), cptr->GetAvailColumns());
+      }
+
+      // Update the work queue and the in-flight cluster list with new requests. We already hold the work queue
+      // mutex
+      // TODO(jblomer): we should ensure that clusterId is given first to the I/O thread.  That is usually the
+      // case but it's not ensured by the code
+      for (const auto &kv : provide) {
+         R__ASSERT(!kv.second.empty());
+
+         RWorkItem workItem;
+         workItem.fClusterId = kv.first;
+         workItem.fColumns = kv.second;
+
+         RInFlightCluster inFlightCluster;
+         inFlightCluster.fClusterId = kv.first;
+         inFlightCluster.fColumns = kv.second;
+         inFlightCluster.fFuture = workItem.fPromise.get_future();
+         fInFlightClusters.emplace_back(std::move(inFlightCluster));
+
+         fWorkQueue.emplace(std::move(workItem));
+      }
+      if (fWorkQueue.size() > 0)
+         fCvHasWork.notify_one();
+   } // work queue lock guard
+
+   return WaitFor(clusterId, columns);
+}
+
+
+ROOT::Experimental::Detail::RCluster *
+ROOT::Experimental::Detail::RClusterPool::WaitFor(
+   DescriptorId_t clusterId, const RPageSource::ColumnSet_t &columns)
+{
+   while (true) {
+      // Fast exit: the cluster happens to be already present in the cache pool
+      auto result = FindInPool(clusterId);
+      if (result) {
+         bool hasMissingColumn = false;
+         for (auto cid : columns) {
+            if (result->ContainsColumn(cid))
+               continue;
+
+            hasMissingColumn = true;
+            break;
+         }
+         if (!hasMissingColumn)
+            return result;
+      }
+
+      // Otherwise the missing data must have been triggered for loading by now, so block and wait
+      decltype(fInFlightClusters)::iterator itr;
+      {
+         std::lock_guard<std::mutex> lockGuardInFlightClusters(fLockWorkQueue);
+         itr = fInFlightClusters.begin();
+         for (; itr != fInFlightClusters.end(); ++itr) {
+            if (itr->fClusterId == clusterId)
+               break;
+         }
+         R__ASSERT(itr != fInFlightClusters.end());
+      }
+
+      auto cptr = itr->fFuture.get();
+      if (result) {
+         result->Adopt(std::move(*cptr));
+      } else {
+         auto idxFreeSlot = FindFreeSlot();
+         fPool[idxFreeSlot] = std::move(cptr);
+      }
+
+      std::lock_guard<std::mutex> lockGuardInFlightClusters(fLockWorkQueue);
```

Comment:
```itr``` is an iterator inside the collection.  Between line 310+ (setting/acquisition) and here (use), the lock is not held by this thread, consequently another thread might be updating the collection in between.  So questions:

a) could the iterator have been invalidated.
b) is it still pointing to the intended element?

So should line 318 and 327 be inside the lock_guard line 309 (and line [319,324] still outside)?

Reply:
It should be fine as the structure of the vector is not modified in the other thread.  I added a clarifying comment:

    // Note that the fInFlightClusters is accessed concurrently only by the I/O thread.  The I/O thread
    // never changes the structure of the in-flight clusters array (it does not add, remove, or swap elements).
    // Therefore, it is safe to access the element pointed to by itr here even after fLockWorkQueue
    // is released.

Reply:
I see.  Conversely the I/O thread is holding the lock for a long time/number-of-lines as it should.

Maybe we ought to comment here why releasing the lock and re-acquiring 'soon' afterward is a gain.

Reply:
I added a clarifying sentence to the comment: `We need to release the lock before potentially blocking on the cluster future.`